redox/redox.go

```diff
@@ -0,0 +1,399 @@
+	handler := &Handler{
+		messagesCollection:                     db.Collection(messagesCollectionName),
+		rescheduledSummaryAndReportsCollection: db.Collection(summaryAndReportsRescheduledOrdersCollectionName),
+		config:                                 config,
+		logger:                                 logger,
+
+		clinics:  clinics,
+		patients: patients,
+	}
+
+	lifecycle.Append(fx.Hook{
+		OnStart: func(ctx context.Context) error {
+			return handler.Initialize(ctx)
+		},
+	})
+
+	return handler, nil
+}
+
+type Handler struct {
+	config                                 Config
+	messagesCollection                     *mongo.Collection
+	rescheduledSummaryAndReportsCollection *mongo.Collection
+	logger                                 *zap.SugaredLogger
+
+	clinics  clinics.Service
+	patients patients.Service
+}
+
+func (h *Handler) Initialize(ctx context.Context) error {
+	_, err := h.messagesCollection.Indexes().CreateMany(ctx, []mongo.IndexModel{
+		{
+			Keys: bson.D{
+				{Key: "meta.Logs.Id", Value: 1},
+			},
+			Options: options.Index().
+				SetName("MetadataLogsId"),
+		},
+		{
+			Keys: bson.D{
+				{Key: "meta.Source.Id", Value: 1},
+				{Key: "meta.FacilityCode", Value: 1},
+			},
+			Options: options.Index().
+				SetName("MetadataSource"),
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	_, err = h.rescheduledSummaryAndReportsCollection.Indexes().CreateMany(ctx, []mongo.IndexModel{
+		{
+			Keys: bson.D{
+				{Key: "createdTime", Value: 1},
+			},
+			Options: options.Index().
+				SetExpireAfterSeconds(int32(rescheduledMessagesExpiration.Seconds())).
+				SetName("CleanupExpiredRescheduledOrders"),
+		},
+	})
+
+	return err
+}
+
+func (h *Handler) VerifyEndpoint(request VerificationRequest) (*VerificationResponse, error) {
+	if request.VerificationToken != h.config.VerificationToken {
+		return nil, fmt.Errorf("%w: invalid verification token", errors.Unauthorized)
+	}
+
+	return &VerificationResponse{Challenge: request.Challenge}, nil
+}
+
+func (h *Handler) AuthorizeRequest(req *http.Request) error {
+	verificationToken := req.Header.Get(verificationTokenHeader)
+
+	if verificationToken == "" {
+		return fmt.Errorf("%w: verification token is required", errors.Unauthorized)
+	}
+	if verificationToken != h.config.VerificationToken {
+		return fmt.Errorf("%w: invalid verification token", errors.Unauthorized)
+	}
+
+	return nil
+}
+
+func (h *Handler) ProcessEHRMessage(ctx context.Context, raw []byte) error {
+	// Deserialize request metadata, so it's easier to process it later
+	message := struct {
+		Meta models.Meta
+	}{}
+	if err := json.Unmarshal(raw, &message); err != nil {
+		return err
+	}
+
+	if !message.Meta.IsValid() {
+		h.logger.Errorw(
+			"unable to process message with invalid metadata",
+			"metadata", message.Meta,
+		)
+		return fmt.Errorf("%w: cannot process message with invalid metadata", errors.BadRequest)
+	}
+
+	bsonRaw := bson.Raw{}
+	if err := bson.UnmarshalExtJSON(raw, true, &bsonRaw); err != nil {
```

Comment:
Are we sure this is safe? bson is a superset of json, so it wont crash, but types may not be as expected? String dates, etc? or is there no non-string fields we want out of this?

Reply:
Makes sense to me.